File: src/main/java/org/academy/api/client/gui/animation/ObjectAnimator.java
package org.academy.api.client.gui.animation;

import org.jspecify.annotations.Nullable;

import java.util.function.Consumer;
import java.util.function.Supplier;

public class ObjectAnimator extends ValueAnimator {
    @Nullable
    private Supplier<Float> getter;
    private boolean autoUpdateStartValue = false;

    private ObjectAnimator() {
    }

    public static ObjectAnimator ofFloat(Consumer<Float> target, float startValue, float endValue) {
        var anim = new ObjectAnimator();
        anim.setFloatValues(startValue, endValue);
        anim.addUpdateListener(animation -> target.accept(animation.getAnimatedValue()));
        target.accept(startValue);
        return anim;
    }

    public static ObjectAnimator ofFloat(Supplier<Float> getter, Consumer<Float> setter, float endValue) {
        var anim = new ObjectAnimator();
        anim.getter = getter;
        anim.setFloatValues(getter.get(), endValue);
        anim.addUpdateListener(animation -> setter.accept(animation.getAnimatedValue()));
        anim.autoUpdateStartValue = true;
        return anim;
    }

    @Override
    public void start() {
        if (autoUpdateStartValue && getter != null) {
            setFloatValues(getter.get(), getEndValue());
        }
        super.start();
    }

    @Override
    void onStartInternal() {
        if (autoUpdateStartValue && getter != null) {
            setFloatValues(getter.get(), getEndValue());
        }
        super.onStartInternal();
    }

    @Override
    public ObjectAnimator setDuration(long duration) {
        super.setDuration(duration);
        return this;
    }

    @Override
    public ObjectAnimator setStartDelay(long startDelay) {
        super.setStartDelay(startDelay);
        return this;
    }

    @Override
    public ObjectAnimator setInterpolator(TimeInterpolator interpolator) {
        super.setInterpolator(interpolator);
        return this;
    }
}

File: src/main/kotlin/org/academy/api/client/gui/animation/ObjectAnimator.kt
package org.academy.api.client.gui.animation

import java.util.function.Consumer

class ObjectAnimator private constructor() : ValueAnimator() {
    private var getter: (() -> Float)? = null
    private var autoUpdateStartValue = false

    override fun start() {
        val getter = getter
        if (autoUpdateStartValue && getter != null) {
            setFloatValues(getter.invoke(), endValue)
        }
        super.start()
    }

    override fun onStartInternal() {
        val getter = getter
        if (autoUpdateStartValue && getter != null) {
            setFloatValues(getter.invoke(), endValue)
        }
        super.onStartInternal()
    }

    override fun setDuration(duration: Long): ObjectAnimator {
        super.setDuration(duration)
        return this
    }

    override fun setStartDelay(startDelay: Long): ObjectAnimator {
        super.setStartDelay(startDelay)
        return this
    }

    override fun setInterpolator(interpolator: TimeInterpolator): ObjectAnimator {
        super.setInterpolator(interpolator)
        return this
    }

    companion object {
        fun ofFloat(target: (Float) -> Unit, startValue: Float, endValue: Float): ObjectAnimator {
            val anim = ObjectAnimator()
            anim.setFloatValues(startValue, endValue)
            anim.addUpdateListener { animation -> target.invoke(animation.animatedValue) }
            target.invoke(startValue)
            return anim
        }

        fun ofFloat(getter: () -> Float, setter: Consumer<Float>, endValue: Float): ObjectAnimator {
            val anim = ObjectAnimator()
            anim.getter = getter
            anim.setFloatValues(getter.invoke(), endValue)
            anim.addUpdateListener { animation -> setter.accept(animation.animatedValue) }
            anim.autoUpdateStartValue = true
            return anim
        }
    }
}